reaktive-annotations/src/commonMain/kotlin/com/badoo/reaktive/annotations/Annotations.kt

```diff
@@ -0,0 +1,43 @@
+package com.badoo.reaktive.annotations
+
+/**
+ * Indicates that return value should normally be used
+ */
+@Retention(AnnotationRetention.SOURCE)
+@Target(AnnotationTarget.FUNCTION)
+annotation class UseReturnValue
+
+/**
+ * Indicates that annotated source does not operate on any particular scheduler
+ */
+@Retention(AnnotationRetention.SOURCE)
+@Target(AnnotationTarget.FUNCTION, AnnotationTarget.FIELD, AnnotationTarget.PROPERTY)
+annotation class EventsOnUnspecifiedScheduler
```

Comment:
Why? RxJava implementation allow us to use any scheduler for scheduled operations (interval, debounce and so on). You want to lock some particular operators to particular scheduler?

Reply:
This kinda we should discuss, perhaps. The idea is that you can annotate a method or a field in e.g. a data source so you clearly see on which thread it emits. So you can decide should you switch scheduler to e.g. main or not. Also maybe we can even introduce some lint checks later.

Use case:
```kotlin
interface UserPersistentDataSource {
    @EventsOnIoScheduler
    fun load(userId: String): Single<User?>
}

interface UserNetworkDataSource {
    @EventsOnMainScheduler
    fun load(userId: String): Single<User>
}

@EventsOnMainScheduler
fun loadUser(userId: String): Single<User> =
    userPersistentDataSource
        .load(userId)
        .flatMap {
            // IO thread here
            it
                ?.toSingle()
                ?.observeOn(mainScheduler) // Need to switch to main thread
                ?: userNetworkDataSource.load(userId) // No need to switch, will be on main thread
        }
```

Reply:
@lukaville @mk-jones Guys please comment your thoughts, this is kinda we already have in Badoo, I find it useful

Reply:
The idea is fine, but we have 2-way control flow for threads. What will happen when `UserPersistentDataSource.load(id).subscribeOn(mainScheduler)` appears in chain? Nobody knows, but consumers will expect results on io thread. Probably it will make sense with ability to freeze subscription on particular thread like `subscribeOnlyOn(ioScheduler)` or `verifyScheduler(ioScheduler)`. Don't know if it is possible to implement.

Reply:
Nope, you are as a developer of data source (or any other component) can annotate its methods with these annotations depending on actual implementations. If you always produce items on a particular thread (you have `observeOn` or `subscribeOn` in your implementation) then you can annotate. If you have neither `observeOn` nor `subscribeOn` then probably `EventsOnUnspecifiedScheduler` would be useful. And so on.
In case of `UserPersistentDataSource.load(id).subscribeOn(mainScheduler)`, you should use 'subscribeOn(ioScheduler)` in your implementation so you always will work on IO thread and `subcribeOn(mainScheduler)` does not matter.

Reply:
So e.g. the following function can be annotated (it's scheduler is guaranteed):
```kotlin
@EventsOnIoScheduler
fun loadUser(): Single<User> =
        single<User> {
            // Load user from database
        }
            .subscribeOn(ioScheduler)

```

Reply:
I like this. It's convenient to know for the api client. Though I would have called it EmitsOnXXX

Reply:
> I like this. It's convenient to know for the api client. Though I would have called it EmitsOnXXX

I also would love "EmitsOnXxx" but it's not really applicable to `Completable`, as it does not emit. But it is still useful in cases: `Completable.andThen(...)`, `Maybe.switchIfEmpty(...)`, etc.

Reply:
➕ 
let's add these annotations